### PR TITLE
Adds elements: Lightweight shared components/Global components

### PR DESCRIPTION
**Pull request by ninjanomnom**
## About The Pull Request

If you came here thinking this was some game feature then you are in the wrong place. Here is where I ramble about code.

This adds `/datum/element` as a sort of sibling to components. Only one of each type gets instanced and they do not get tied directly to any particular thing like a component does. Basically they're a very lightweight component for doing simple functionality that doesn't have much state.

Originally this concept came about as a kind of component that could be shared between many parents to reduce some resource costs. Doing this would allow us to componentize more behaviors that are a part of too many things to be viable to have a whole component for every single one. For example a component on every space turf would be entirely unviable. With elements it's much more reasonable.

This implements a prety bare framework and a couple components are migrated to it. It's ready to be used but I fully expect I'm going to need to refine how it works for all the usecases we'll want it for.

Also: this fixes the qdeleted signal. This signal isn't even possible because after qdel is done there's nothing to receive a signal anyway. I've changed it to a qdeling signal instead. I need it to work for some elements to know when to clean themselves up.

### CODE REVIEW

**Comment by Qustinnus** [comment]
Only the Avatar, master of all four elements, could stop them, but when the world needed him most, he vanished.

**Comment by AnturK** [comment]
Why not simple subtype ? As far as i can tell it would only need special handling in AddComponent ?

**Comment by ninjanomnom** [comment]
They don't actually need to share any functionality with components and there are a lot of things I would need to handle specially.

File: code/__DEFINES/components.dm [hunk 1]
```diff
@@ -31,7 +36,7 @@
 #define COMSIG_COMPONENT_ADDED "component_added"				//when a component is added to a datum: (/datum/component)
 #define COMSIG_COMPONENT_REMOVING "component_removing"			//before a component is removed from a datum because of RemoveComponent: (/datum/component)
 #define COMSIG_PARENT_PREQDELETED "parent_preqdeleted"			//before a datum's Destroy() is called: (force), returning a nonzero value will cancel the qdel operation
-#define COMSIG_PARENT_QDELETED "parent_qdeleted"				//after a datum's Destroy() is called: (force, qdel_hint), at this point none of the other components chose to interrupt qdel and Destroy has been called
+#define COMSIG_PARENT_QDELETING "parent_qdeleted"				//after a datum's Destroy() is called: (force), at this point none of the other components chose to interrupt qdel and Destroy has been called
```
Comment (SpaceManiac) [inline code/__DEFINES/components.dm]: should change the value, docs, and comment in garbage.dm to match the new behavior

File: code/datums/elements/_element.dm [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+/datum/proc/AddElement(eletype, ...)
+	var/datum/element/ele = SSdcs.GetElement(eletype)
+	args[1] = src
+	if(ele.Attach(arglist(args)) == ELEMENT_INCOMPATIBLE)
+		CRASH("Incompatible [eletype] assigned to a [type]! args: [json_encode(args)]")
+
+/datum/proc/RemoveElement(eletype)
+	var/datum/element/ele = SSdcs.GetElement(eletype)
+	ele.Detach(src)
```
Comment (SpaceManiac) [inline code/datums/elements/_element.dm]: I'm not seeing any evidence that this actually unregisters signals

**Review by SpaceManiac — DISMISSED** [review]
Any estimate of how much memory usage this saves?

File: code/datums/elements/_element.dm (hunk 2 above)
Comment (ninjanomnom) [reply code/datums/elements/_element.dm]: Why would it?

Comment (SpaceManiac) [reply code/datums/elements/_element.dm]: You add the cleaning element to something, then remove it, but it keeps cleaning

Comment (ninjanomnom) [reply code/datums/elements/_element.dm]: Oh you mean that particular element should, yeah this RemoveElement proc was an after thought as nothing actually removes them in what I was testing

Comment (SpaceManiac) [reply code/datums/elements/_element.dm]: It should either be removed or work correctly

**Comment by ninjanomnom** [comment]
I have no clue how much this saves memory wise and neither of these examples are used very extensively. This is more to make some concepts that aren't viable with components because of their memory costs possible.

**Comment by SpaceManiac** [comment]
What did you have in mind?

**Comment by SpaceManiac** [comment]
Merge conflict

**Pull request merged**